### PR TITLE
Kubeadm graduate kubeconfig phase

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR graduates phase kubeconfig the kubeadm init workflow.

Following subcommands are automatically added to kubeadm init by the PhaseRunner:

```
kubeadm init phase kubeconfig
kubeadm init phase kubeconfig admin
kubeadm init phase kubeconfig controller-manager
kubeadm init phase kubeconfig kubelet
kubeadm init phase kubeconfig scheduler
```

Following alpha phase sub command are removed

```
kubeadm alpha phase kubeconfig
kubeadm alpha phase kubeconfig admin
kubeadm alpha phase kubeconfig controller-manager
kubeadm alpha phase kubeconfig kubelet
kubeadm alpha phase kubeconfig scheduler
```

**Special notes for your reviewer: **:

WIP waiting for other phases to graduate.
Please consider only latest two commits.

**Which issue(s) this PR fixes** : 
Rif https://github.com/kubernetes/kubeadm/issues/1163

/sig cluster-lifecycle
/kind feature
/cc @neolit123 
/assign @timothysc 
@kubernetes/sig-cluster-lifecycle-pr-reviews 
@chuckha 
@rosti 
@ereslibre 

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Review by neolit123 — APPROVED** [review]
the last two commits (kubeconfig) LGTM.
will double check them again before merge.
thanks.

File: cmd/kubeadm/app/cmd/phases/kubeconfig.go [hunk 1]
```diff
@@ -19,77 +19,135 @@ package phases
...
+			short: "Generates a kubeconfig file for the scheduler to use",
+			long:  "Generates the kubeconfig file for the scheduler to use and saves it to %s file.",
+		},
+	}
+)
 
-		Please note that this should *only* be used for bootstrapping purposes. After your control plane is up,
-		you should request all kubelet credentials from the CSR API.
-		`+cmdutil.AlphaDisclaimer), filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.KubeletKubeConfigFileName))
+// kubeConfigData defines the behavior that a runtime data struct passed to the kubeconfigt phase
```
Comment (yagonobre) [inline cmd/kubeadm/app/cmd/phases/kubeconfig.go]: typo `kubeconfigt`

**Comment by fabriziopandini** [comment]
@timothysc @neolit123 now also this one should be ready 
I included in the PR some small fix/improvements to existing phases to make things more consistent

File: cmd/kubeadm/app/cmd/phases/certs.go [hunk 2]
```diff
@@ -217,10 +248,24 @@ func runCertPhase(cert *certsphase.KubeadmCert, caCert *certsphase.KubeadmCert)
...
 
+		// if external CA mode, skip certificate generation
+		if data.ExternalCA() {
+			fmt.Printf("[certs] External CA mode: Using existing %s certificate\n", cert.BaseName)
+			return nil
+		}
+
+		// if using external etcd, skips etcd certificates generation
+		if data.Cfg().Etcd.External != nil && cert.CAName == "etcd-ca" {
+			fmt.Printf("[certs] External ETCD mode: Skipping %s certificate authority generation\n", cert.BaseName)
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/certs.go]: ETCD lowercase -> etcd.

File: cmd/kubeadm/app/cmd/phases/certs.go [hunk 3]
```diff
@@ -195,17 +181,62 @@ func getSANDescription(certSpec *certsphase.KubeadmCert) string {
...
 
+		// if external CA mode, skips certificate authority generation
+		if data.ExternalCA() {
+			fmt.Printf("[certs] External CA mode: Using existing %s certificate authority\n", ca.BaseName)
+			return nil
+		}
+
+		// if using external etcd, skips etcd certificate authority generation
+		if data.Cfg().Etcd.External != nil && ca.Name == "etcd-ca" {
+			fmt.Printf("[certs] External ETCD mode: Skipping %s certificate authority generation\n", ca.BaseName)
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/certs.go]: ETCD lowercase -> etcd.

File: cmd/kubeadm/app/cmd/phases/kubeconfig.go [hunk 4]
```diff
@@ -19,77 +19,150 @@ package phases
...
 	kubeconfigphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/kubeconfig"
 	"k8s.io/kubernetes/pkg/util/normalizer"
 )
 
 var (
-	allKubeconfigLongDesc = normalizer.LongDesc(`
-		Generates all kubeconfig files necessary to establish the control plane and the admin kubeconfig file.
-		` + cmdutil.AlphaDisclaimer)
+	kubeconfigLongDesc = normalizer.LongDesc(`
+	Kubeconfig file utilities.
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/kubeconfig.go]: Kubeconfig lowercase -> kubeconfig

File: cmd/kubeadm/app/cmd/phases/kubeconfig.go [hunk 5]
```diff
@@ -19,77 +19,150 @@ package phases
...
+	}
+}
 
 // NewCmdKubeConfig returns main command for kubeconfig phase
 func NewCmdKubeConfig(out io.Writer) *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "kubeconfig",
-		Short: "Generates all kubeconfig files necessary to establish the control plane and the admin kubeconfig file",
-		Long:  cmdutil.MacroCommandLongDescription,
+		Short: "Kubeconfig file utilities",
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/kubeconfig.go]: Kubeconfig lowercase.

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go [hunk 6]
```diff
@@ -248,12 +225,12 @@ func createKubeConfigFileIfNotExists(outDir, filename string, config *clientcmda
 
 	// Check if the file exist, and if it doesn't, just write it to disk
 	if _, err := os.Stat(kubeConfigFilePath); os.IsNotExist(err) {
+		fmt.Printf("[kubeconfig] Writing %q KubeConfig file\n", filename)
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: KubeConfig lowercase -> kubeconfig

File: cmd/kubeadm/app/cmd/phases/certs.go [hunk 7]
```diff
@@ -71,6 +70,7 @@ var (
 // (and thus with different runtime data struct, all of them requested to be compliant to this interface)
 type certsData interface {
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/certs.go]: btw, unless i'm mistaken allCertsLongDesc and allCertsExample are unused in this file.

File: cmd/kubeadm/app/cmd/phases/certs.go [hunk 8]
```diff
@@ -195,17 +181,62 @@ func getSANDescription(certSpec *certsphase.KubeadmCert) string {
...
+	defer func() { cfg.CertificatesDir = data.CertificateDir() }()
+
+	// create the new service account key (or use existing)
+	return certsphase.CreateServiceAccountKeyAndPublicKeyFiles(cfg)
+}
+
+func runCerts(c workflow.RunData) error {
+	data, ok := c.(kubeConfigData)
+	if !ok {
+		return errors.New("kubeconfigFile phase invoked with an invalid data struct")
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/certs.go]: "kubeconfigFile phase" -> "certs phase"  ?

File: cmd/kubeadm/app/cmd/phases/kubeconfig.go [hunk 9]
```diff
@@ -19,77 +19,150 @@ package phases
...
 
-	userKubeconfigLongDesc = normalizer.LongDesc(`
-		Outputs a kubeconfig file for an additional user.
-		` + cmdutil.AlphaDisclaimer)
+// runKubeConfigFile executes kubeconfig creation logic.
+func runKubeConfigFile(kubeConfigFileName string) func(workflow.RunData) error {
+	return func(c workflow.RunData) error {
+		data, ok := c.(kubeConfigData)
+		if !ok {
+			return errors.New("kubeconfigFile phase invoked with an invalid data struct")
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/kubeconfig.go]: perhaps use `kubeconfig`, as `kubeconfigFile` is not a real phase name?
optionally use `kubeConfigFileName` to clarify?

File: cmd/kubeadm/app/cmd/phases/kubeconfig.go [hunk 10]
```diff
@@ -19,77 +19,150 @@ package phases
...
+	}
+}
 
-	controllerManagerKubeconfigLongDesc = fmt.Sprintf(normalizer.LongDesc(`
-		Generates the kubeconfig file for the controller manager to use and saves it to %s file.
-		`+cmdutil.AlphaDisclaimer), filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.ControllerManagerKubeConfigFileName))
+func runKubeConfig(c workflow.RunData) error {
+	data, ok := c.(kubeConfigData)
+	if !ok {
+		return errors.New("kubeconfigFile phase invoked with an invalid data struct")
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/kubeconfig.go]: kubeconfigFile -> kubeconfig

**Review by neolit123 — COMMENTED** [review]
@fabriziopandini thanks for the updated added some minor comments about variable names and stdout text. :+1:

File: cmd/kubeadm/app/phases/certs/certs.go [hunk 11]
```diff
@@ -261,18 +258,19 @@ func writeKeyFilesIfNotExist(pkiDir string, baseName string, key *rsa.PrivateKey
 		// kubeadm doesn't validate the existing certificate key more than this;
 		// Basically, if we find a key file with the same path kubeadm thinks those files
 		// are equal and doesn't bother writing a new file
-		fmt.Printf("[certificates] Using the existing %s key.\n", baseName)
+		fmt.Printf("[certs] Using the existing %q key\n", baseName)
 	} else {
 
 		// Write .key and .pub files to disk
+		fmt.Printf("[certs] Generated %q key and public key\n", baseName)
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/certs.go]: Generated -> Generating

File: cmd/kubeadm/app/cmd/phases/kubeconfig.go [hunk 12]
```diff
@@ -19,77 +19,150 @@ package phases
...
 	"k8s.io/kubernetes/pkg/util/normalizer"
 )
 
 var (
-	allKubeconfigLongDesc = normalizer.LongDesc(`
-		Generates all kubeconfig files necessary to establish the control plane and the admin kubeconfig file.
-		` + cmdutil.AlphaDisclaimer)
+	kubeconfigLongDesc = normalizer.LongDesc(`
+	Kubeconfig file utilities.
+	` + cmdutil.AlphaDisclaimer)
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/phases/kubeconfig.go]: AlphaDisclaimer?

Comment (timothysc) [reply cmd/kubeadm/app/cmd/phases/kubeconfig.go]: What's the split here?  We can follow up on another PR if needed.

File: cmd/kubeadm/app/cmd/phases/kubeconfig.go [hunk 13]
```diff
@@ -19,77 +19,150 @@ package phases
...
+		}
 
-	userKubeconfigExample = normalizer.Examples(`
-		# Outputs a kubeconfig file for an additional user named foo
-		kubeadm alpha phase kubeconfig user --client-name=foo
-		`)
-)
+		// if external CA mode, skip certificate authority generation
+		if data.ExternalCA() {
+			//TODO: implement validation of existing kubeconfig files
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/phases/kubeconfig.go]: We should convert TODOs to issues

**Review by timothysc — APPROVED** [review]
@fabriziopandini thank you for the patch! 

Comments are none blocking, and we can follow on. 
In the future please do refactor/cleanup work in a separate PRs. 

/lgtm 
/approve

File: cmd/kubeadm/app/cmd/phases/kubeconfig.go (hunk 10 above)
Comment (timothysc) [reply cmd/kubeadm/app/cmd/phases/kubeconfig.go]: [kubeconfig] ?

File: cmd/kubeadm/app/cmd/phases/kubeconfig.go (hunk 9 above)
Comment (timothysc) [reply cmd/kubeadm/app/cmd/phases/kubeconfig.go]: I think having the print [phase] = error [phase] should probably match.

File: cmd/kubeadm/app/cmd/phases/certs.go [hunk 14]
```diff
@@ -195,17 +181,62 @@ func getSANDescription(certSpec *certsphase.KubeadmCert) string {
...
+	// if external CA mode, skip service account key generation
+	if data.ExternalCA() {
+		fmt.Printf("[certs] External CA mode: Using existing sa keys\n")
+		return nil
+	}
+
+	// if dryrunning, write certificates from a temporary folder
+	cfg := data.Cfg()
+	cfg.CertificatesDir = data.CertificateWriteDir()
+	defer func() { cfg.CertificatesDir = data.CertificateDir() }()
```
[the remark this answers is not in the document]
Comment (ereslibre) [reply cmd/kubeadm/app/cmd/phases/certs.go]: The idea behind this was to unset the `CertificatesDir` change, even if the phase logic grows more complex and we have several exit points in the future. Maybe it's less clear at the moment that we only have one exit point...

File: cmd/kubeadm/app/cmd/phases/certs.go (hunk 7 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/certs.go]: done

File: cmd/kubeadm/app/cmd/phases/certs.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/certs.go]: @timothysc yes, the logic around paths and temporary paths could be improved.
There is already an open issue, but definitely +100 on code audit sessions

File: cmd/kubeadm/app/cmd/phases/kubeconfig.go (hunk 12 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/kubeconfig.go]: This is for the bit that is not strictly a phase but should be moved on kubeadm alpha (without phases)
I have already a PR in flight to move those bits under `cmd\alpha`

File: cmd/kubeadm/app/cmd/phases/kubeconfig.go (hunk 13 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/kubeconfig.go]: +1

**Review by neolit123 — APPROVED** [review]
LGTM.

File: cmd/kubeadm/app/cmd/phases/certs.go [hunk 15]
```diff
@@ -195,17 +164,62 @@ func getSANDescription(certSpec *certsphase.KubeadmCert) string {
...
+
+	// create the new service account key (or use existing)
+	return certsphase.CreateServiceAccountKeyAndPublicKeyFiles(cfg)
+}
+
+func runCerts(c workflow.RunData) error {
+	data, ok := c.(kubeConfigData)
```
Comment (ereslibre) [inline cmd/kubeadm/app/cmd/phases/certs.go]: Is this `kubeConfigData` in purpose instead of a `certsData`?

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve

**Pull request merged**